F-FrontEnd/src/run_tests.py

```diff
@@ -0,0 +1,782 @@
+                            test_case = args[1]
+                            test_case_num = test_case_num_by_name[test_case]
+                            results[test_case_num] = (test_case, res)
+                            if not res.result:
+                                num_failed_testcases += 1
+                            if print_progress:
+                                print('[%s of %s] %s' % (num_finished_testcases, num_test_cases, test_case))
+                                if not res.result:
+                                    print(res.text_summary())
+                    if num_failed_testcases > 0:
+                        return RC.FAILURE, results
+            else:
+                test_case = self.args.test_case
+                if print_progress:
+                    print(test_case)
+                res = run_test_case(working_dir, test_case, self.args, testcase_deps[test_case], False,
+                                    working_dir_is_tmp)
+                results = [(test_case, res)]
+                if not res.result:
+                    if print_progress:
+                        print(res.text_summary())
+                    return RC.FAILURE, results
+        return RC.SUCCESS, results
+
+    def run(self, print_progress=False) -> TestResult:
+        start_ts = dt.datetime.utcnow()
+        return_code, results = self.__run(print_progress)
+        end_ts = dt.datetime.utcnow()
+        res = TestResult(testcases=tuple(results),
+                         result=return_code == RC.SUCCESS,
+                         start_timestamp=start_ts, end_timestamp=end_ts)
+        if print_progress:
+            print('Elapsed time: ', end_ts - start_ts)
+        if self.args.report_file is not None:
+            res.save_to_file(self.args.report_file)
+        return res
+
+    @staticmethod
+    def scan_for_dependencies(dir_path: str,
+                              debug_output: bool = False) -> Tuple[Tuple[str, ...], Dict[str, Tuple[str, ...]]]:
+        test_cases = []
+        reg_comment = re.compile('!.*')
+        spaces = '[\s]+'
+        opt_spaces = '[\s]*'
+        fortran_id = '[a-z][a-z0-9\_]*'
+        module_decl = 'module' + spaces + ('(%s)' % fortran_id)
+        submodule_name = '[\sa-z0-9\_]+'
+        submodule_decl = 'submodule' + opt_spaces + ('\((%s)\)' % submodule_name) + opt_spaces + ('(%s)' % fortran_id)
+        reg_space = re.compile('\s')
+        reg_module_decl = re.compile(module_decl)
+        reg_submodule_decl = re.compile(submodule_decl)
+        reg_use_module = re.compile('use' + spaces + ('(%s)' % fortran_id))
+        mod_to_file = {}
+        testcase_deps = {}
+        for pattern in ('*.f', '*.f90', '*.f08'):
+            test_cases += [os.path.basename(str(path)) for path in pathlib.Path(dir_path).glob(pattern)]
+        test_cases = sorted(test_cases)
+        # Order of the tests is currently relevant, as some files depend on each other!!!
```

Comment:
So basically, alphabetical order implies that the testcases are topologically sorted?
Otherwise, the code below looks buggy to me. Might be a nice opportunity to remove that requirement?

Reply:
Alas, some test cases have repeating module names, so this ordering had to be preserved. Initially I wanted to switch to explicit dependencies, but it was too much work to setup.
This one will have to wait for another time.